Source/GR32_Math.pas

```diff
@@ -1224,13 +1224,26 @@ function Sign(Value: Integer): Integer;
 {$IFDEF FPC} assembler; nostackframe; {$ENDIF}
 asm
 {$IFDEF TARGET_x64}
-        MOV     EAX, Value
-{$ENDIF}
+  {$IFDEF MSWINDOWS}

```

Comment:
Looks good - But why not use the same for 32-bit?

Reply:
This was due to careful cycle counting.  Under 32-bit, the cycle count is about 3, assuming NEG can execute at the same time as CDQ.  Under 64-bit, the cycle count is at least 4 due to the need to move the formal parameter into EAX first.

For the new version, XOR and TEST can execute at the same time (1 cycle apiece) as they work on different registers, while SETG depends on both of these instructions completing, while SAR ECX, 31 can execute at the same time, before the results are merged into the LEA instruction for 3 cycles overall.  From my own experimentation in the past, if you zero the whole register, there's no partial write penalty from using SETcc on the mapped byte register.

Long story short, the reason why the 32-bit vrersion wasn't changed is because there's no actual speed saving, but the existing algorithm takes fewer bytes of machine code at only 5 bytes (ultimately it probably won't matter since code pointers are aligned to at least a 16-byte granularity).

Reply:
> Under 64-bit, the cycle count is at least 4 due to the need to move the formal parameter into EAX first.

What move? You mean the LEA?
Apart from that I got the same timings.

| Op | Latency |
| ---- | ---- |
| XOR | 0/1 |
| TEST | 1 |
| SETG | 1 |
| SAR | 1 | 
| LEA |  1 |

| Op | Latency |
| ---- | ---- |
| CDQ | 1 |
| NEG | 1 |
| ADC | 1 |
| MOV | 1 |

Reply:
I meant the ```MOV EAX, Value``` instruction that has to appear before ```CDQ``` on 64-bit, which will add at least 1 to the latency.

Reply:
In the meantime I added an extra comment for the 32-bit version to explain why it's still using the old algorithm.